bulldozer/signals.go

```diff
@@ -21,62 +21,186 @@ import (
 
 	"github.com/palantir/bulldozer/pull"
+	"github.com/pkg/errors"
 	"github.com/rs/zerolog"
 )
 
+type Signal interface {
+	Enabled() bool
+	// Determine if the signal has values assigned to it and should be considered when matching
+
+	Matches(context.Context, pull.Context, string) (bool, string, error)
+	// Determine if the signal matches a value in the target pull request
+}
+
+type Labels []string
+type CommentSubstrings []string
+type Comments []string
+type PRBodySubstrings []string
+type Branches []string
+type BranchPatterns []string
+type MaxCommits int
+
 type Signals struct {
-	Labels            []string `yaml:"labels"`
-	CommentSubstrings []string `yaml:"comment_substrings"`
-	Comments          []string `yaml:"comments"`
-	PRBodySubstrings  []string `yaml:"pr_body_substrings"`
-	Branches          []string `yaml:"branches"`
-	BranchPatterns    []string `yaml:"branch_patterns"`
+	Labels            `yaml:"labels"`
+	CommentSubstrings `yaml:"comment_substrings"`
+	Comments          `yaml:"comments"`
+	PRBodySubstrings  `yaml:"pr_body_substrings"`
+	Branches          `yaml:"branches"`
+	BranchPatterns    `yaml:"branch_patterns"`
+	MaxCommits        `yaml:"max_commits"`
+}
+
+func (signal *Labels) Enabled() bool {
+	return len(*signal) > 0
+}
+
+func (signal *CommentSubstrings) Enabled() bool {
+	return len(*signal) > 0
+}
+
+func (signal *Comments) Enabled() bool {
+	return len(*signal) > 0
+}
+
+func (signal *PRBodySubstrings) Enabled() bool {
+	return len(*signal) > 0
+}
+
+func (signal *Branches) Enabled() bool {
+	return len(*signal) > 0
+}
+
+func (signal *BranchPatterns) Enabled() bool {
+	return len(*signal) > 0
+}
+
+func (signal *MaxCommits) Enabled() bool {
+	return *signal > 0
 }
 
 func (s *Signals) Enabled() bool {
-	size := 0
-	size += len(s.Labels)
-	size += len(s.CommentSubstrings)
-	size += len(s.Comments)
-	size += len(s.PRBodySubstrings)
-	size += len(s.Branches)
-	size += len(s.BranchPatterns)
-	return size > 0
+	return s.Labels.Enabled() ||
+		s.CommentSubstrings.Enabled() ||
+		s.Comments.Enabled() ||
+		s.PRBodySubstrings.Enabled() ||
+		s.Branches.Enabled() ||
+		s.BranchPatterns.Enabled() ||
+		s.MaxCommits.Enabled()
 }
 
-// Matches returns true if the pull request meets one or more signals. It also
+// MatchesAll returns true if the pull request matches ALL of the signals. It also
+// returns a description of the match status. The tag argument appears
+// in this description and indicates the behavior (trigger, ignore) this
+// set of signals is associated with.
+func (s *Signals) MatchesAll(ctx context.Context, pullCtx pull.Context, tag string) (bool, string, error) {
+	signals := []Signal{
+		&s.Labels,
+		&s.CommentSubstrings,
+		&s.Comments,
+		&s.PRBodySubstrings,
+		&s.Branches,
+		&s.BranchPatterns,
+		&s.MaxCommits,
+	}
+
+	for _, signal := range signals {
+		matches, _, err := signal.Matches(ctx, pullCtx, tag)
+		if err != nil {
+			return false, "", err
+		}
+
+		if signal.Enabled() && !matches {
```

Comment:
I see you added debug logging for when `Matches` is called but the signal is disabled. I'm worried that could get quite verbose for conditions that only use one or two signals. It might be better to make this `Enabled` check before calling `Matches` and then skip the signal if it is not enabled.